ws-server/src/server.rs

```diff
@@ -309,19 +315,42 @@ async fn background_task(
 	let stop_server2 = stop_server.clone();
 	let sink = MethodSink::new_with_limit(tx, max_response_body_size);
 
+	let mut submit_ping = tokio::time::interval(ping_interval);
+	let ping_submitted_tx = Arc::new(AtomicBool::new(false));
+	let ping_submitted_rx = ping_submitted_tx.clone();
+
 	middleware.on_connect();
 
 	// Send results back to the client.
 	tokio::spawn(async move {
 		while !stop_server2.shutdown_requested() {
-			if let Some(response) = rx.next().await {
-				// If websocket message send fail then terminate the connection.
-				if let Err(err) = send_ws_message(&mut sender, response).await {
-					tracing::warn!("WS send error: {}; terminate connection", err);
-					break;
+			select! {
+				rx_value = rx.next() => {
+					if let Some(response) = rx_value {
+						// If websocket message send fail then terminate the connection.
+						if let Err(err) = send_ws_message(&mut sender, response).await {
+							tracing::warn!("WS send error: {}; terminate connection", err);
+							break;
+						}
+					} else {
+						break;
+					}
+				}
+
+				_ = submit_ping.tick() => {
+					// Ping was already submitted.
+					if ping_submitted_tx.load(Ordering::Relaxed) {
```

Comment:
looks to me that we should use stricter ordering here,

as `store's` from can re-ordered arbitrary with `Relaxed`, so I think it's possible that the will be set "to the wrong value" i.e, even if the ping was sent it could be overwritten.

Please use `release-acquire or seqcst` for this unless you can prove that I'm wrong :)